src/databricks/labs/dqx/check_funcs.py

```diff
@@ -1784,31 +1786,62 @@ def has_valid_schema(
             - A closure that applies the schema check and adds the necessary condition columns.
 
     Raises:
-        InvalidParameterError: If the schema string is invalid or cannot be parsed, or if
-             the input schema is neither a string nor a StructType.
+        InvalidParameterError:
+            - If the *expected_schema* string is invalid or cannot be parsed
+            - If *expected_schema* is neither a string nor a StructType
+            - If both *expected_schema* and *ref_table* are specified
+            - If neither *expected_schema* nor *ref_table* are specified
+
+    Note:
+        Exactly one of *expected_schema* or *ref_table* must be specified.
     """
+    if expected_schema and ref_table:
+        raise InvalidParameterError(
+            "Cannot specify both 'expected_schema' and 'ref_table' when using 'has_valid_schema'"
+        )
+
+    if not expected_schema and not ref_table:
+        raise InvalidParameterError(
+            "Must specify one of 'expected_schema' or 'ref_table' when using 'has_valid_schema'"
+        )
 
     column_names: list[str] | None = None
     if columns:
         column_names = [get_column_name_or_alias(col) if not isinstance(col, str) else col for col in columns]
 
-    _expected_schema = _get_schema(expected_schema, column_names)
+    if expected_schema:
+        expected_schema_struct = _get_schema(expected_schema, column_names)
+
     unique_str = uuid.uuid4().hex  # make sure any column added to the dataframe is unique
     condition_col = f"__schema_condition_{unique_str}"
     message_col = f"__schema_message_{unique_str}"
 
-    def apply(df: DataFrame) -> DataFrame:
+    def apply(df: DataFrame, spark: SparkSession) -> DataFrame:
```

Comment:
we should make it compatible with other methods that take reference table. There should be an option to provide the reference data frame as well similar to foreign_key check.